### PR TITLE
Implement the 2 new enum modes on PaddingMode

### PR DESCRIPTION
**Pull request by AlexGhiondea**
This is the code port **only**, I am working on tests next,

### CODE REVIEW

File: src/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs [hunk 1]
```diff
@@ -50,6 +51,29 @@ private byte[] PadBlock(byte[] block, int offset, int count)
...
+                    break;
+
+                // ISO padding fills the blocks up with random bytes and adds the total number of padding
+                // bytes as the last pad byte, adding an extra block if the last block is complete.
+                //
+                // xx rr rr rr rr rr rr 07
+                case PaddingMode.ISO10126:
+                    result = new byte[count + padBytes];
+                    
+                    //CapiNative.UnsafeNativeMethods.CryptGenRandom(m_provider, result.Length - 1, result);
```
Comment (bartonjs) [inline src/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs]: Delete commented-out code.

File: src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs [hunk 2]
```diff
@@ -157,6 +157,34 @@ private byte[] DepadBlock(byte[] block, int offset, int count)
             // See PadBlock for a description of the padding modes.
             switch (PaddingMode)
             {
+                case PaddingMode.ANSIX923:
+                    padBytes = block[offset + count - 1];
+
+                    // Verify the amount of padding is reasonable
+                    if (padBytes <= 0 || padBytes > InputBlockSize) {
```
Comment (bartonjs) [inline src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs]: Style: Wrong bracing is used here. (Expected: Allman, Actual: K&R)

File: src/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs [hunk 3]
```diff
@@ -50,6 +51,28 @@ private byte[] PadBlock(byte[] block, int offset, int count)
                     Buffer.BlockCopy(block, offset, result, 0, result.Length);
                     break;
 
+                // ANSI padding fills the blocks with zeros and adds the total number of padding bytes as
+                // the last pad byte, adding an extra block if the last block is complete.
+                //
+                // x 00 00 00 00 00 00 07
+                case PaddingMode.ANSIX923:
+                    result = new byte[count + padBytes];
+                    Buffer.BlockCopy(block, 0, result, 0, count);
```
Comment (AlexGhiondea) [inline src/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs]: `Buffer.BlockCopy(block, 0, result, 0, count);` 
should be
`Buffer.BlockCopy(block, offset, result, 0, count);`

and same for the other one.

**Comment by AlexGhiondea** [comment]
/cc @steveharter @bartonjs I have  updated the code and added a couple of tests to ensure we can roundtrip with the new padding mode (and one to make sure we don't :) ).

I could not find a good way to get a hold of the decryption buffer before the `DepadBlock` method was called. That would have allowed me to make sure the right padding is actually happening for the right padding mode. If you have any ideas how to do that cleanly, let me know and I will!

File: src/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs [hunk 4]
```diff
@@ -18,6 +18,7 @@ namespace Internal.Cryptography
     //
     internal sealed class UniversalCryptoEncryptor : UniversalCryptoTransform
     {
+        private static readonly RandomNumberGenerator s_randomNumberGenerator = RandomNumberGenerator.Create();
         public UniversalCryptoEncryptor(PaddingMode paddingMode, BasicSymmetricCipher basicSymmetricCipher)
```
Comment (bartonjs) [inline src/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs]: A blank line here would be nice :)

File: src/System.Security.Cryptography.Algorithms/tests/System.Security.Cryptography.Algorithms.Tests.csproj [hunk 5]
```diff
@@ -142,7 +144,7 @@
     <Compile Include="TripleDesTests.cs" />
     <Compile Include="$(CommonPath)\Interop\Unix\Interop.Libraries.cs">
       <Link>Common\Interop\Unix\Interop.Libraries.cs</Link>
-    </Compile>
+    </Compile>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/System.Security.Cryptography.Algorithms.Tests.csproj]: I'm a little confused by the trailing space addition here.  (It also feels like the whole edit to this file could be reverted, since it just moved a line up... though it did group it with its friends...)

File: src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs [hunk 6]
```diff
@@ -136,10 +136,19 @@ private bool DepaddingRequired
...
+                {
+                    case PaddingMode.PKCS7:
+                    case PaddingMode.ANSIX923:
+                    case PaddingMode.ISO10126:
+                        return true;
+                    case PaddingMode.Zeros:
+                    case PaddingMode.None:
+                        return false;
+                    default:
+                        throw new CryptographicException(SR.Cryptography_InvalidPadding);
```
Comment (bartonjs) [inline src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs]: Consider something like `Debug.Fail($"Unknown padding mode {PaddingMode}");`, so the type of padding that was missing is more easily identified.

Though, really... you'd need to be adding a new one... so... probably doesn't matter?

**Comment by bartonjs** [comment]
>  could not find a good way to get a hold of the decryption buffer before the DepadBlock method was called. That would have allowed me to make sure the right padding is actually happening for the right padding mode. If you have any ideas how to do that cleanly, let me know and I will!

If I were adding padding tests right now, what I'd do is
* Foreach padding mode p != None
  * Foreach size in (0, 3, 7, blocksize)
    * Encrypt size bytes with padding p
      * Verify blocksize == size added a block (p!=Zeros), otherwise only one block out
    * Decrypt with Padding None (which effectively returns the buffer prior to DepadBlock :smile:)
      * Verify decrypt[last]
      * Verify other things (PKCS7 and ANSI are semantic, if ISO has more than 2 padding bytes at least one filler byte should be nonzero)

For depad testing
* Foreach size in (0, 3, 7, blocksize)
  * Foreach padding mode p != None
  * Build a block manually to conform to p (two blocks for size==blocksize,p!=Zeros)
  * Encrypt with Padding None
  * Decrypt with padding p
  * Then, do a positional tamper for each filler byte and ensure the expected behaviors
    * Zeros, ISO won't care, won't throw.
    * PKCS7 and ANSI will care, will throw.

Where "filler byte" means "padding byte other than the last one".
Oh, and do that for 3DES (BlockSize=8 bytes) and AES (BlockSize=16 bytes).

If you don't want to do all that, we could just create a test enhancement bug and let it get addressed later.

File: src/System.Security.Cryptography.Algorithms/tests/System.Security.Cryptography.Algorithms.Tests.csproj (hunk 5 above)
Comment (AlexGhiondea) [inline src/System.Security.Cryptography.Algorithms/tests/System.Security.Cryptography.Algorithms.Tests.csproj]: This is due to the merge with master. There was a build break I had to fix locally (adding the package reference) but it was also fixed in master and I missed this when I merged.

It makes sense to undo the changes to this file.

File: src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs (hunk 6 above)
Comment (AlexGhiondea) [inline src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs]: Seeing how long it took to identify this in the first place, I will add the debug call :).

Even if we don't add new padding modes, the overhead for the debug call is minimal.

**Comment by AlexGhiondea** [comment]
@bartonjs I forgot about `PaddingMode.None` :). 

I am going to try and add a couple more tests around this.

Any other comments on the PR?

**Comment by bartonjs** [comment]
> Any other comments on the PR?

Nah, just the Debug.Fail and the newline.  So we'll see how the tests go :smile:

**Comment by AlexGhiondea** [comment]
@bartonjs added some tests around padding. It covers some of the scenarios you mentioned. We should create a separate item to add even more testing here.

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs [hunk 7]
```diff
@@ -324,5 +324,115 @@ public static void ZeroLengthTransformPKCS7()
                 }
             }
         }
+
+#if netstandard17
```
Comment (joperezr) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: Should we move this to a different file and make it a partial class? That is the norm we have been following so far.

File: src/System.Security.Cryptography.Cng/src/System.Security.Cryptography.Cng.csproj [hunk 8]
```diff
@@ -261,6 +261,9 @@
     <ProjectReference Include="..\..\System.Runtime.Extensions\pkg\System.Runtime.Extensions.pkgproj" />
     <!-- ToDo: Remove this P2P reference once new packages are produced and updated -->
     <ProjectReference Include="..\..\System.IO\pkg\System.IO.pkgproj" />
```
Comment (joperezr) [inline src/System.Security.Cryptography.Cng/src/System.Security.Cryptography.Cng.csproj]: Can you remove these other P2Ps? corefx versions were updated this morning so I believe we should be able to clean this up.

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/TripleDES/TripleDESCipherTests.cs [hunk 9]
```diff
@@ -90,6 +90,80 @@ public static void TripleDESRoundTrip192BitsZerosECB()
             }
         }
 
+#if netstandard17
```
Comment (AlexGhiondea) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/TripleDES/TripleDESCipherTests.cs]: In the crypto tests we have been going the #if-def approach.

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs (hunk 7 above)
Comment (AlexGhiondea) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: For crypto we have been using the #if-def pattern so far.

File: src/System.Security.Cryptography.Cng/src/System.Security.Cryptography.Cng.csproj (hunk 8 above)
Comment (AlexGhiondea) [inline src/System.Security.Cryptography.Cng/src/System.Security.Cryptography.Cng.csproj]: I think they can be removed. I will file an issue to remove all of them at once.

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs (hunk 7 above)
Comment (bartonjs) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: Since these tests are focused highly on the padding (instead of other AES-ness) these could make sense in a different file.  We've just frequently (in crypto) had an obvious "this belongs with these other tests, and we wouldn't have thought of putting it in a different file except for the 1.6 test build".

I'm fine with whichever way.

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs [hunk 10]
```diff
@@ -324,5 +324,115 @@ public static void ZeroLengthTransformPKCS7()
...
+        [Fact]
+        public static void PaddingMode_Validation()
+        {
+            byte[] key = "1ed2f625c187b993256a8b3ccf9dcbfa5b44b4795c731012f70e4e64732efd5d".HexToByteArray();
+            byte[] iv = "47d1e060ba3c8643f9f8b65feeda4b30".HexToByteArray();
```
Comment (bartonjs) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: Key and IV seem unused here.

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs [hunk 11]
```diff
@@ -324,5 +324,115 @@ public static void ZeroLengthTransformPKCS7()
...
+            plainText = "e505a2".HexToByteArray();
+            ValidatePaddingMode(plainText, "0a2e62938b03e5822ee251117a4ce066".HexToByteArray(), PaddingMode.Zeros, 13);
+            ValidatePaddingMode(plainText, "46785bde46622b92ff7c8ebb91508a4d".HexToByteArray(), PaddingMode.PKCS7, 13);
+            ValidatePaddingMode(plainText, "43b27d41a9fde73ca5db22c0fda76cb1".HexToByteArray(), PaddingMode.ANSIX923, 13);
+            ValidatePaddingMode(plainText, Array.Empty<byte>(), PaddingMode.ISO10126, 13);
+
+            plainText = "46785bde46622b92ff7c8ebb91508a".HexToByteArray();
+            ValidatePaddingMode(plainText, "1be8aa365a15d11fc7826b3a10602d09".HexToByteArray(), PaddingMode.Zeros, 1);
+            ValidatePaddingMode(plainText, "db5b7829cce732bfe609140cf45a8843".HexToByteArray(), PaddingMode.PKCS7, 1);
+            ValidatePaddingMode(plainText, "db5b7829cce732bfe609140cf45a8843".HexToByteArray(), PaddingMode.ANSIX923, 1);
```
Comment (bartonjs) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: Can we get these as [Theory] on the ValidatePaddingMode method?

```c#
[Theory]
[InlineData(PaddingMode.Zeros, 1, "46785bde46622b92ff7c8ebb91508a", "1be8aa365a15d11fc7826b3a10602d09")]
[InlineData(PaddingMode.PKCS7, 1, "46785bde46622b92ff7c8ebb91508a", "db5b7829cce732bfe609140cf45a8843")]
[InlineData(PaddingMode.ANSIX923, 1, "46785bde46622b92ff7c8ebb91508a", null)]
```

(Yeah, I switched the argument order to make the test case name start off with the relevant parts)

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs [hunk 12]
```diff
@@ -324,5 +324,115 @@ public static void ZeroLengthTransformPKCS7()
...
+            for (int i = buffer.Length - expectedPaddingSize; i < buffer.Length - 1; i++)
+            {
+                Assert.Equal(0, buffer[i]);
+            }
+        }
+
+        private static void ValidateISO10126Padding(byte[] buffer, int expectedPaddingSize)
+        {
+            // there is nothing else to validate as all the other padding bytes are random.
+            Assert.Equal(buffer[buffer.Length - 1], expectedPaddingSize);
```
Comment (bartonjs) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: We *could* assert that there's at least one non-zero value among the rest of the padding.  The likelihood of that failing is 1/(256^(expectedPaddingSize-1)).  So if expectedPaddingSize is at least 5 there's down to a 1 in 4 billion chance of random failure.  (So we could only do the check if expectedPaddingSize is >= 5)

**Comment by AlexGhiondea** [comment]
@bartonjs I filed an issue for this and will send a PR shortly :).

File: src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs (hunk 12 above)
Comment (AlexGhiondea) [inline src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: So you are saying randomly pick one of the bytes and check if it is non-zero? 
I can add it, but I don't think it will add too much value. I submitted #13902 -- we can chat more there.
